File: 3077. Maximum Strength of K Disjoint Subarrays.py
#PYTHON
#*3077. Maximum Strength of K Disjoint Subarrays
class Solution:
    def maximumStrength(self, nums, k):
        dp = [[-float('inf')] * len(nums) for _ in range(k)]
        dp[k - 1][0] = nums[0] * k

        for i in range(1, len(nums)):
            for j in range(k - 1, -1, -1):
                dp[j][i] = max(dp[j][i - 1], (dp[j + 1][i - 1] if j + 1 < k else 0)) + nums[i] * (j + 1) * (-1 if j & 1 == 1 else 1)

        return max(dp[0])
    
s=Solution()
print(s.maximumStrength([1,2,3,-1,2], 3), 22) 
print(s.maximumStrength([12,-2,-2,-2,-2], 5), 64) 
print(s.maximumStrength([-1, -2, -3], 1), -1) 
print(s.maximumStrength([-99, 85], 1), 85)
print(s.maximumStrength([2,-76,77,-30,5,-96,24], 3), 497)

File: 3077. Maximum Strength of K Disjoint Subarrays.js
//3077. Maximum Strength of K Disjoint Subarrays
nums = [1,2,3,-1,2]
k = 3
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maximumStrength = function(nums, k) {
  var dp0 = Array(k).fill().map(() => Array(nums.length).fill(-Infinity));
  var dp1 = Array(k).fill().map(() => Array(nums.length).fill(-Infinity));
  dp1[k - 1][0] = nums[0] * k;

  for (var i = 1; i < nums.length; ++i) {
    for (var j = k - 1; j >= 0; --j) {
      dp0[j][i] = Math.max(dp0[j][i - 1], dp1[j][i - 1]);
      dp1[j][i] = Math.max(dp1[j][i - 1], dp0[j + 1]?.[i] || 0) + nums[i] * (j + 1) * (j & 1 ? -1 : 1);
    }
  }

  return Math.max(...dp1[0]);
};
console.log(maximumStrength())